File: 01-OOP-Refresher/src/go/dog.go
package main

import "fmt"

type Dog struct {
	Name  string
	Breed string
}

func (d *Dog) GetName() string { return d.Name }

func (d *Dog) Eat() {
	fmt.Printf("%s is eating.\n", d.Name)
}

func (d *Dog) MakeSound() {
	fmt.Printf("%s says Woof! [%s]\n", d.Name, d.Breed)
}

func (d *Dog) Fetch() {
	fmt.Printf("%s fetches the ball!\n", d.Name)
}

File: 01-OOP-Refresher/src/rust/dog.rs
use crate::animal::Animal;

pub struct Dog {
    name: String,
    breed: String,
}

impl Dog {
    pub fn new(name: &str, breed: &str) -> Self {
        Dog {
            name: name.to_string(),
            breed: breed.to_string(),
        }
    }

    pub fn fetch(&self) {
        println!("{} fetches the ball!", self.name);
    }
}

impl Animal for Dog {
    fn name(&self) -> &str {
        &self.name
    }

    fn make_sound(&self) {
        println!("{} says Woof! [{}]", self.name, self.breed);
    }
}
